http/transactions_middleware.go

```diff
@@ -0,0 +1,110 @@
+// it attempts to resolve (commit or rollback) the transaction on the context of the request
+// if the transaction failed for some reason it catches the error and returns a 500 response
+// with a clear error indicating the failure.
+type TransactionResponseWriter struct {
+	http.ResponseWriter
+	request        *http.Request
+	transactionErr error
+	headerWritten  bool
+}
+
+// Header returns the https response header.
+func (t *TransactionResponseWriter) Header() http.Header {
+	return t.ResponseWriter.Header()
+}
+
+// WriteHeader writes the status code to the response.
+// In this implementation we DO NOT write the actual status code to the
+// underlying header to avoid multiple response.WriteHeader calls error
+// but rather we save the status code and write it back only if the transaction
+// resolved successfully.
+func (t *TransactionResponseWriter) WriteHeader(statusCode int) {
+	// TODO: Note that this approach of resolving the transaction when the application has
+	// started writing the response isn't generally correct. It works because all our handlers
+	// currently complete the database activities before starting to write the response. But
+	// that may change and then we would potentially be committing unfinished transactions.
+	t.headerWritten = true
+	err := transaction.Resolve(t.request.Context())
+	if err != nil {
+		// Keeping the error to return to the user - see `Write`.
+		t.transactionErr = err
+		t.ResponseWriter.Header().Set("Content-Type", "application/json")
+		t.ResponseWriter.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	t.ResponseWriter.WriteHeader(statusCode)
+}
+
+func (t *TransactionResponseWriter) Write(payload []byte) (int, error) {
+	// Some applications may start writing the response body before calling the `WriteHeader`
+	// method. That is completely OK according to the documentation. But if we don't call it
+	// explicitly the transaction will never be resolved and the connection will be leaked.
+	if !t.headerWritten {
+		t.WriteHeader(http.StatusOK)
+	}
+	if t.transactionErr != nil {
+		body := E500.Format(t.request, "An error occurred: %v", t.transactionErr)
+		data, err := json.Marshal(body)
+		// The following 2 cases return size actually written,
+		// which may be surprising to caller that expected len(payload).
+		// But it's simplest to not mislead logging/middleware about actual size of response.
+		if err != nil {
+			return t.ResponseWriter.Write(panicBody)
+		}
+		return t.ResponseWriter.Write(data)
+	}
+	return t.ResponseWriter.Write(payload)
+}
+
+func (t *TransactionResponseWriter) WriteErrIfNothingWritten(r *http.Request) {
+	if !t.headerWritten {
+		body := E500.Format(r, "")
+		SendError(t.ResponseWriter, r, &body)
+	}
+}
+
+// TransactionMiddleware creates a new HTTP middleware that begins a database transaction and stores
+// it in the request context.
+func TransactionMiddleware(next http.Handler) http.Handler {
```

Comment:
This "middleware" should use the same style used by other HTTP handler wrappers in this package. It should be named `HandlerWrapper` have a builder to configure it. I'd expect to use it like this:

```go
// Create the transaction manager:
manager, err := transaction.NewManager().
        Logger(logger).
        Handle(handle).
        Build(ctx)
if err != nil {
        ...
}

// Create the transaction handler wrapper:
wrapper, err := transaction.NewHandlerWrapper().
        Logger(logger).
        Manager(manager).
        Build(ctx)
if err != nil {
        ...
}

// Wrap the handler:
wrappedHandler := wrapper.Wrap(originalHandler)
```